composer/datasets/c4.py

```diff
@@ -0,0 +1,333 @@
+                                                                  mlm_probability=self.mlm_probability)
+        # Return DataSpec
+        return DataSpec(dataloader=dataloader_hparams.initialize_object(
+            dataset=c4_dataset,
+            batch_size=batch_size,
+            sampler=None,
+            drop_last=self.drop_last,
+            collate_fn=collate_fn,
+        ),
+                        split_batch=_split_dict_fn)
+
+
+class C4Dataset(IterableDataset):
+    """Builds a streaming, sharded, sized IterableDataset for the C4 (Colossal Cleaned CommonCrawl) dataset. Used for
+    pretraining autoregressive or masked language models. Text samples are streamed directly from the cloud using
+    HuggingFace's C4 Dataset with streaming backend (See https://huggingface.co/datasets/c4 for more details). The text
+    samples are then shuffled, tokenized, and grouped on-the-fly.
+
+    Parameters:
+        split (str): What split of the dataset to use. Either `train` or `validation`
+        max_samples (int): Max number of post-processed token samples, used to set epoch size of the IterableDataset.
+        tokenizer_name (str): The name of the HuggingFace tokenizer to preprocess text with.
+        max_seq_len (int): The max sequence length of each token sample.
+        group_method (str): How to group text samples into token samples. Either `truncate` or `concat`.
+        shuffle (bool): Whether to shuffle the samples in the dataset. Currently, shards are assigned deterministically to devices, but shuffling affects the order of shards (per-device), and the order of samples via (per-device) shuffle buffers (Default: `False`)
+        shuffle_buffer_size (int): If `shuffle=True`, samples are read into a buffer of this size (per-device), and randomly sampled from there to produce shuffled samples (Default: `10000`)
+        seed (int): If `shuffle=True`, what seed to use for shuffling operations. (Default: `5`)
+    Returns:
+        A :class:`torch.utils.data.IterableDataset` object
+    """
+
+    def __init__(self,
+                 split,
+                 max_samples,
+                 tokenizer_name,
+                 max_seq_len,
+                 group_method,
+                 shuffle=False,
+                 shuffle_buffer_size=10000,
+                 seed=5):
+        try:
+            import datasets
+            import transformers
+        except ImportError:
+            raise ImportError('HuggingFace transformers and datasets are not installed. '
+                              'Please install with `pip install composer[nlp]`')
+
+        self.split = split
+        self.max_samples = max_samples
+        self.tokenizer_name = tokenizer_name
+        self.max_seq_len = max_seq_len
+        self.group_method = group_method
+        self.shuffle = shuffle
+        self.shuffle_buffer_size = shuffle_buffer_size
+        self.seed = seed
+
+        # Metadata
```

Comment:
just to check, this is specific for C4 correct?

can you encode it in a single `dict`? e.g.
```
c4_metadata = {"train": {"num_shards" = 1024, .. }, "validation": {"num_shards": 8, ..}}
```

Reply:
Yes, this is very specific to the C4 dataset built/uploaded by AllenAI and hosted on HuggingFace. I can simplify it to a dict.